### DISCUSSION TITLE
picking up our discussion on distillery/brewery use

### DISCUSSION THREAD

**Discussion by bsmokeman** [Brewken/brewken#28] [General]
Hey Matt Young, 
I'm just now starting to dig through code, looking at where to start for my project, and was interested in what appears to be a lot of work on replacing the database layer which is the primary starting point I needed to accomplish.
Do you have a quick summary of the changes made to the initial brewtarget code, and are you keeping your code in sync with brewtarget in other areas, or is this a branch that will completely go its own direction?
I'd like to join forces and branch, and implement the changes I need in order to tie in batch management, and centralized database, where the gui just launches as a client for a centralized server database storage.  Multiple clients could then run simultaneously in different locations within the distillery/brewery.  I'll be implementing a data sync with our ERP software, which I plan to use dollibar.
I'm trying to achieve a completely open source run distillery/brewery.
Thanks,
Barry Smoke
barrylsmoke@gmail.com

**Comment by matty0ung** [comment]
Hi Barry,

I've tried to keep Brewken in sync with Brewtarget, picking up fixes from Brewtarget and offering back my own enhancements and fixes.  At some point, I expect there will be features in Brewken that Brewtarget does not want because they're not relevant to homebrewers, but my hope is always to keep the recipe side of things in sync.

The new database layer is, as you've seen, implemented in Brewken and I made a PR (https://github.com/Brewtarget/brewtarget/pull/617) to merge it back to Brewtarget which helped find & fix some bugs in the new code.

That Brewtarget PR should give you an idea of the scope of the change (ie quite big).  Look at comments in `database/ObjectStore.h` for how the new layer works.  Then `database/ObjectStoreTyped.cpp` holds the mapping from object property names to/from database table and column names.  If you look at `model/NamedEntity.h` and `model/NamedEntity.cpp` you'll see code that allows object setters to push changes down to the database layer without too much boilerplate.

I think doing a centralised database would be easier now, albeit still non-trivial.  ObjectStore can already talk to a remote database, so I would imagine needing some mechanism for it to periodically query whether the DB has been modified (by other clients), and then to propagate those changes into the objects stored in memory (plus update the UI).  Might also need some lock mechanism to prevent two people trying to edit the same recipe etc at the same time. Would be good to talk through this in more detail.
Best,
Matt

**Reply by bsmokeman** [reply]
yes, I could see it being very difficult to keep in sync, after more thought about how I would want to store, and organize things differently in order to keep track of them, compared to current functionality. Perhaps a limitation to core brew functionality, but I'm starting to see a lot of divergence. Everything alcohol related starts with a mash of some kind, even though those mashes are treated differently through their fermentation life, or have a conversion done for brewed beverages.  I can see wanting to start there, with a mash screen, then that mash actually move through its process to a whiskey derivative for further tracking pre-distillation, or a fruit brandy section.  A fruit brandy could also become a fruit wine if not distilled.  Perhaps even both a fruit brandy, and a fruit wine, splitting the mash up in purpose.  I also see the need to filter based on month/year.  The database will contain all created batches through their history, and a distillery and brewery such as I envision will have so many batches, big batches, small batches, ciders, etc, that displaying historical data will be difficult on processing and need to be limited to a current working set.  I'll also need some reporting, but I haven't decided if that will be on the brew software side, or the ERP/dolibarr side of the business.  I'm still working this out.  
I'd be interested in your thoughts of where you are headed, and how you are using this for your brewery in south-west France.  Are some of the things I've stated overkill for your use scenario? As I see the notes, 'not compileable yet', you might not be, or you might be using brewtarget as is currently, and hoping to migrate to this.  That's what I was considering doing at first. I have a few months to see what I can accomplish during our construction phase, so I'm diving in.  Once we get funded, and I have a base direction for where I'm headed with the project, I may throw a little development money at it as well.

**Reply by matty0ung** [reply]
Brewken does compile - on Linux, Mac and Windows.  It does everything Brewtarget does, so there's no reason not to use it.  If you point me at "not compileable yet" notes, I'll correct them! :smiley: 

My first question would be how important it is to support both distilleries and beer-making.  It feels like it would add a lot of complexity - not least to the UI - to be so generic.  This is partly because the process to make distilled drinks and beers is a bit different, and partly because they are different types of business.  Beer is much faster turn-around and much higher volume than whiskey or brandy, I would think.  (Similarly, although I have fun comparing notes with one of my neighbours who runs a small vineyard, it feels like there are also a lot of differences between producing wine and beer.)  Would we risk being "jack of all trades" in trying to support so many different needs?

If there really is a strong need to support multiple different types of drink (brandy / whiskey / fruit wine / beer / mead / etc) then I'd think we'd want to be able to have a beer version of the UI that hides a lot of details that brewers generally don't need to worry about.  (Brewing is complicated enough as it is! :smile:)   What's your background though? Are you a distiller?

For me, getting started with a small microbrewery, it seemed there wasn't any software that bridged the gap between enthusiastic homebrewer and small-scale licenced professional brewer.  I now have to worry about paying excise duty, stock control (of brewed beers and ingredients), pricing, sales tracking, and all sorts of other things.  I spent a while making spreadsheets to get going, but that will become painful as we grow, so ... software to the rescue.  At least that's the theory.

In practice, the Brewtarget base on which we're building has needed a bit of bringing up to date, so I haven't yet got on to a lot of the new functionality.  (Though I am now using the improved alcohol-by-volume calculation for my Customs declarations!)

**Reply by bsmokeman** [reply]
cool, I'll find where I saw that, and just start using your code base.
also cool you are starting to use it for actual work.  
My background is a 28+ year IT professional, web developer/programmer, and also drinker and alcohol enthusiast who through other business ventures, and employment failures decided to do this distillery/brewery venture as my last business venture pre-retirement.  A lot of craft distilleries and breweries specialize in one or the other, with only a fraction across the globe doing both.  if you google brewstillery you will find an ample amount doing both though.  I'm not the first.  I initially wanted to do it for use of my waste products, making beverages out of distillery spent mash.  The process is going to be complicated to prepare that spent wort for consumption, and I'm not quite sure about that part yet, but I'm still preparing to make and experiment with all possible alcoholic products my 2 permits will allow.  Here in Arkansas, USA, the distillery, and brewery have to be separately permitted.  The craft distillery permit allows all possible combinations of mash using arkansas grown products.  The brewery has no limitations on products combined for brewed beverages, although I will include arkansas products into those as well.
self-customization of open source to accomplish this goal is exactly why I'm using this software as a starting point.
I would think a mode selection could be incorporated, so we were developing on the same gui, and in a preferences screen, we could have 'include distillery functionality', 'include cidery functionality' that could simply hide those subdivisions of the tabbed gui that's already in place.  I look for simple solutions where ever possible.
I agree on finding software that bridges the gap of a small microbrewery instead of homebrew software.  The minimum equipment I'm looking at going with for this venture is the brewzilla b150, their largest brewing vessel.  I will have multiple B150's, either combining them during the brew process for large batches, sparged and pumped together simultaneously during the boil, or do individual seperate batches of different flavors, as I get an idea of which beers my local market has demand for.  I also think its my unique way of thinking about batches and reportability/accountability that will make this possible.  You also have to understand that brewtarget/brewken is only brewing software, and not an MRP(Material Requirements Planning), or ERP(Enterprise Resource Planning) software.  I chose dollibarr because it is one of the only ERP software packages out that is open source.  Its also written in php, which I have a lot of experience in.  MRP isn't functional enough for local and federal tracking, and reporting. So, dolibarr is actually going to be the software that runs our enterprise/venture.  I learned about dolibarr when I looked for an ERP software to track and generate computer server sku's for manufacturing.  With a sync program for dolibarr, we can import resultant batches from brewken, even getting bulk liquid resultant amounts, and then account for the bottling, print trackable labels(which will tie back into the originating batch) 

I am concerned about the whole open source vs commercial aspect.  I believe under our chosen licenses so far for keeping the projects open source, the code could not be packaged for other distilleries to use.  A discussion for later, once functionality is achieved.

also remember my last goal is to 'package' each batch for automation to a controller, and my target so far is to also go open source with that solution using raspberry pi with either craftbeerpi, or brewpi.  haven't gotten that far yet.
I'm thinking at first those would just be steps to follow along, and manually trigger the elements, following boil timers, etc...but eventually would be a completely automated process with manual confirmation things were going ok via confirm screens.
lots more to come...

**Reply by bsmokeman** [reply]
on the licenses, its actually QT, not the project code I'm talking about there.  just one aspect to think about.  The QT commercial licenses are really expensive.

**Reply by bsmokeman** [reply]
saw where I saw the compile comment, its on the doc folder, so probably just a documentation comment

**Reply by matty0ung** [reply]
Open source isn't necessarily incompatible with doing something commercially.  A good proportion of clients might be happy to pay some small monthly subscription for the centralised part of the software to be hosted for them rather than set it up and maintain it themselves.  (Even just knowing you've always got up-to-date lists of grains, yeasts, hops, style guides and so on would be valuable.)  Plus there may be regulatory advantages to the brewer in this approach - eg I can say to the tax authorities "Everything is recorded centrally on this service I subscribe to and I can't tamper with the data, so you can rely on it as a definitive record of what I brewed and sold, and that it's automatically done all the ABV and weight-to-volume calculations to give me the right figures for my duty declarations".
